File: app/src/main/java/com/larryhsiao/nyx/attachments/NewAttachment.java
package com.larryhsiao.nyx.attachments;

import com.silverhetch.clotho.Source;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

import static java.sql.Statement.RETURN_GENERATED_KEYS;

/**
 * New Attachment of a Jot
 */
public class NewAttachment implements Source<Attachment> {
    private final Source<Connection> source;
    private final String uri;
    private final long jotId;

    public NewAttachment(Source<Connection> source, String uri, long jotId) {
        this.source = source;
        this.uri = uri;
        this.jotId = jotId;
    }

    @Override
    public Attachment value() {
        try (PreparedStatement stmt = source.value().prepareStatement(
            // language=H2
            "INSERT INTO attachments(uri, jot_id) " +
                "VALUES (?,?)",
            RETURN_GENERATED_KEYS
        )) {
            stmt.setString(1, uri);
            stmt.setLong(2, jotId);
            stmt.executeUpdate();
            final ResultSet res = stmt.getGeneratedKeys();
            if (!res.next()) {
                throw new IllegalArgumentException("Creating Attachment failed, jotId: "+ jotId+", Uri: "+ uri);
            }
            return new ConstAttachment(res.getLong(1), jotId, uri);
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: core/src/main/java/com/larryhsiao/nyx/core/attachments/NewAttachment.kt
package com.larryhsiao.nyx.core.attachments

import com.silverhetch.clotho.Source
import java.sql.Connection
import java.sql.Statement

/**
 * New Attachment of a Jot
 */
class NewAttachment(
    private val source: Source<Connection>,
    private val uri: String,
    private val jotId: Long
) : Source<Attachment?> {
    override fun value(): Attachment {
        try {
            source.value().prepareStatement( // language=H2
                "INSERT INTO attachments(uri, jot_id) " +
                    "VALUES (?,?)",
                Statement.RETURN_GENERATED_KEYS
            ).use { stmt ->
                stmt.setString(1, uri)
                stmt.setLong(2, jotId)
                stmt.executeUpdate()
                val res = stmt.generatedKeys
                require(res.next()) { "Creating Attachment failed, jotId: $jotId, Uri: $uri" }
                return ConstAttachment(res.getLong(1),
                    jotId,
                    uri,
                    1,
                    0
                )
            }
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }
}